standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/PartFilterParser.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.metastore.parser;
+
+import java.util.function.Function;
+
+import org.antlr.v4.runtime.*;
+import org.antlr.v4.runtime.misc.Interval;
+import org.antlr.v4.runtime.misc.ParseCancellationException;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PartFilterParser {
+  private static final Logger LOG = LoggerFactory.getLogger(PartFilterParser.class);
+  private static final ThreadLocal<AstBuilder> astBuilder = ThreadLocal.withInitial(AstBuilder::new);
+
+  public static ExpressionTree parseFilter(String filter) throws MetaException {
+    return parse(filter, parser -> astBuilder.get().visitFilter(parser.filter()));
+  }
+
+  private static <T> T parse(String filter, Function<PartitionFilterParser, T> toResult)
+      throws MetaException {
+    LOG.debug("Parsing filter: " + filter);
+    CharStream upperCaseCharStream = new UpperCaseCharStream(CharStreams.fromString(filter));
+    PartitionFilterLexer lexer = new PartitionFilterLexer(upperCaseCharStream);
+    lexer.removeErrorListeners();
+    lexer.addErrorListener(ParseErrorListener.INSTANCE);
+
+    CommonTokenStream tokenStream = new CommonTokenStream(lexer);
+    PartitionFilterParser parser = new PartitionFilterParser(tokenStream);
+    parser.removeErrorListeners();
+    parser.addErrorListener(ParseErrorListener.INSTANCE);
+
+    try {
+      return toResult.apply(parser);
+    } catch (ParseCancellationException e) {
+      throw new MetaException("Error parsing partition filter: " + e.getMessage());
+    }
+  }
+
+  /** Case-insensitive ANTLR string stream */
+  private static class UpperCaseCharStream implements CharStream {
```

Comment:
How about extending `CodePointCharStream` instead of wrapping it?
In that case only `int LA(int i)` would be overriden.

It is done for antlr 3
https://github.com/apache/hive/blob/7cd3107a76d633ef5fae2ffb8ec16953ac968092/parser/src/java/org/apache/hadoop/hive/ql/parse/ANTLRNoCaseStringStream.java#L37-L51